source/solvers/heat_transfer_assemblers.cc

```diff
@@ -523,3 +522,118 @@ HeatTransferAssemblerViscousDissipation<dim>::assemble_rhs(
 
 template class HeatTransferAssemblerViscousDissipation<2>;
 template class HeatTransferAssemblerViscousDissipation<3>;
+
+
+
+template <int dim>
+void
+HeatTransferAssemblerLaser<dim>::assemble_matrix(
+  HeatTransferScratchData<dim> & /*scratch_data*/,
+  StabilizedMethodsCopyData & /*copy_data*/)
+{}
+
+template <int dim>
+void
+HeatTransferAssemblerLaser<dim>::assemble_rhs(
+  HeatTransferScratchData<dim> &scratch_data,
+  StabilizedMethodsCopyData &   copy_data)
+{
+  // Laser parameters
+  const double concentration_factor = laser_parameters->concentration_factor;
+  const double laser_power          = laser_parameters->laser_power;
+  const double absorptivity         = laser_parameters->laser_absorptivity;
+  const double penetration_depth    = laser_parameters->penentration_depth;
+  const double laser_start_time     = laser_parameters->start_time;
+  const double laser_end_time       = laser_parameters->end_time;
+  const double beam_radius          = laser_parameters->beam_radius;
+  const double current_time = this->simulation_control->get_current_time();
+
+  if (current_time >= laser_start_time && current_time <= laser_end_time)
+    {
+      // Get laser path
+      Function<dim> &laser_path = *(laser_parameters->laser_path);
+      laser_path.set_time(current_time);
+
+      // Get laser lcoation
+      Point<dim> laser_location;
+      laser_location[0] =
+        laser_path.value(laser_location,
+                         laser_parameters->perpendicular_plane_coordinate_one);
+      laser_location[1] =
+        laser_path.value(laser_location,
+                         laser_parameters->perpendicular_plane_coordinate_two);
+
+      double laser_location_in_depth = 0.0;
+      if constexpr (dim == 3)
+        {
+          laser_location[2] =
+            laser_path.value(laser_location,
+                             laser_parameters->beam_direction_coordinate);
+          laser_location_in_depth = laser_location[2];
+        }
+
+      // Get laser location on the operation surface
+      Point<2> laser_location_on_surface;
+      laser_location_on_surface[0] = laser_location[0];
+      laser_location_on_surface[1] = laser_location[1];
+
+      const unsigned int n_q_points = scratch_data.n_q_points;
+      const unsigned int n_dofs     = scratch_data.n_dofs;
+
+      // Copy data elements
+      auto &strong_residual = copy_data.strong_residual;
+      auto &local_rhs       = copy_data.local_rhs;
+
+      // assembling right hand side
+      for (unsigned int q = 0; q < n_q_points; ++q)
+        {
+          // Get quadrate point location
+          Point<2> quadrature_point_on_surface;
+          quadrature_point_on_surface[0] = scratch_data.quadrature_points[q][0];
+          quadrature_point_on_surface[1] = scratch_data.quadrature_points[q][1];
```

Comment:
I am not sure how this work in 2D? Should it not be that  the second variable is the depth in 2D? Also, shouldnt the depth depend on the direction / orientation of the laser? This is a bit unclear to me right now.

Reply:
True, it was my mistake. It is corrected now.